### DISCUSSION TITLE
How to use cpu_bound? Code will be called twiced on initial button press

### DISCUSSION THREAD

**Discussion by hengst85** [zauberzeug/nicegui#2596] [Q&A] [answered]
### Question

Hi, I build a GUI with a frame (one class) and two tables (one class for each table) inside. The table data will be computed in separate functions. This functions will be called, when the tables are initialized and when the user push a button (The user can choose whether the entire table or a single row should be updated). Therefore I use the run.cpu_bound methode.  

```python
from nicegui import ui, run
from time import sleep
    
def call_taskA(x):
    print("Tasks A will be executed")
    sleep(1)
    return x
    
def call_taskB(x):
    print("Tasks B will be executed")
    sleep(1)
    return x


class frame():
    def __init__(self) -> None:
        with ui.row().classes('w-full items-center justify-between'):
            self.subclassA = tableA()
            self.subClassB = tableB()        


class tableA():
    def __init__(self) -> None:
        result = call_taskA('a')
        print(f"Task A finished: {result}") 
        ui.button("Update Table A", on_click= lambda: self.run_taskA('a2'))
    
    async def run_taskA(self, x):
        print("Task A will be started")
        result = await run.cpu_bound(call_taskA, x)
        print(f"Task A finished: {result}")
        
        
class tableB():
    def __init__(self) -> None:
        result = call_taskB('b')
        print(f"Task B finished: {result}")
        ui.button("Update Table B", on_click= lambda: self.run_taskB('b2'))

    async def run_taskB(self, x):
        print("Task B will be started")
        result = await run.cpu_bound(call_taskB, x)
        print(f"Task B finished: {result}")


with ui.column().classes('w-full'):
    frame()

ui.run()
```

When I started the GUI, I got following output;

Tasks A will be executed
Task A finished: a
Tasks B will be executed
Task B finished: b

When I pressed one of the buttons the first time, I got the following output:

Task A will be started
Tasks A will be executed
Task A finished: a
Tasks B will be executed
Task B finished: b
Tasks A will be executed
Task A finished: a2

So my classes / tables will be initiliazed once again. Depending on the amount of data, this can take a very long time. 
Then everything runs as desired. How can I work around the fact that the code takes so long for the initial user input?

**Comment by falkoschindler** [comment, marked as answer]
Hi @hengst85,

This is an interesting problem. Let's boil it down to the following example:
```py
class table:
    def __init__(self) -> None:
        print('Init')
        ui.button('Update', on_click=lambda: run.cpu_bound(print, 'Running CPU-bound task'))

table()
```

When clicking the button, we see another "Init". This is because `cpu_bound` needs to create a new process to run the `print` function. In order to get the new process running, it needs to import the Python file, which will call `table()` again.

You can avoid that by moving your UI definition from the global scope into a page function:
```py
@ui.page('/')
def page():
    table()
```

Now the page isn't recreated for the new process and you don't see another "Init".

**Reply by hengst85** [reply]
Hi falkoschindler,

thanks for your quick reply. I tried your approach and changed my example code:

```python
from nicegui import ui, run
from time import sleep
    
def call_taskA(x):
    print("Tasks A will be executed")
    sleep(3)
    return x
    
def call_taskB(x):
    print("Tasks B will be executed")
    sleep(3)
    return x  


class tableA():
    def __init__(self) -> None:
        result = call_taskA('a')
        print(f"Task A finished: {result}") 
        ui.button("TaskA", on_click= lambda: self.run_taskA('a2'))
    
    async def run_taskA(self, x) -> None:
        print("Task A will be started")
        result = await run.cpu_bound(call_taskA, x)
        print(f"Task A finished: {result}")
        
        
class tableB():
    def __init__(self) -> None:
        result = call_taskB('b')
        print(f"Task B finished: {result}")
        ui.button("TaskB", on_click= lambda: self.run_taskB('b2'))

    async def run_taskB(self, x) -> None:
        print("Task B will be started")
        result = await run.cpu_bound(call_taskB, x)
        print(f"Task B finished: {result}")


@ui.page('/')
def frame():
    with ui.row().classes('w-full items-center justify-between'):
        table_a = tableA()
        table_b = tableB() 

ui.run()

```

That looks great to me. I'm going to try this approach for my original GUI as soon as possible. 

A nice side effect is that the GUI automatically refreshes every time the page is opened.

**Reply by hengst85** [reply]
Hi @falkoschindler,

this approach is working as expected. But know now I have another problem:
I add a timer to update the table(s) all 15min and will send a notification to the user (with windows toasts) if something has changed. But the timer is only working, if the page is open in a browser window. 
Is it possible to get the best from both implementations:
- Process is working in the background 
- Send notifications to the user from background or when UI is open
- Tables will be updated when user open the UI in the browser
- No double code running due to cpu_bound using

**Reply by rodja** [reply]
We are working on a new example with "global computation" https://github.com/zauberzeug/nicegui/pull/2685 (work in progress). Maybe it could help you with your new problem/question? Otherwise I would kindly ask you to create a new question to keep topics in order.
